src/compiler/moduleSpecifiers.ts

```diff
@@ -832,17 +847,54 @@ const enum MatchingMode {
     Pattern,
 }
 
-function tryGetModuleNameFromExports(options: CompilerOptions, targetFilePath: string, packageDirectory: string, packageName: string, exports: unknown, conditions: string[], mode = MatchingMode.Exact): { moduleFileToTry: string; } | undefined {
+function getOutputPath(targetFilePath: string, options: CompilerOptions, commonDir: string) {
+    if (options.emitDeclarationOnly) {
+        return;
+    }
+    return changeExtension(
+        options.outDir ?
+            resolvePath(
+                options.outDir,
+                getRelativePathFromDirectory(commonDir, targetFilePath, /*ignoreCase*/ false),
+            ) :
+            targetFilePath,
+        getOutputExtension(targetFilePath, options),
+    );
+}
+
+function tryGetModuleNameFromExportsOrImports(options: CompilerOptions, host: ModuleSpecifierResolutionHost, getCanonicalFileName: GetCanonicalFileName, targetFilePath: string, packageDirectory: string, packageName: string, exports: unknown, conditions: string[], mode: MatchingMode, isImports: boolean): { moduleFileToTry: string; } | undefined {
     if (typeof exports === "string") {
+        const currentDirectory = host.getCurrentDirectory();
+        const commonDir = getNormalizedAbsolutePath(
+            getCommonSourceDirectory(
+                options,
+                () => [],
+                currentDirectory,
+                getCanonicalFileName,
+            ),
+            currentDirectory,
+        );

```

Comment:
I somewhat struggled with finding the right things to use here. I think that I got it right but it's best to review this with caution. I was considering using `info.sourceDirectory` and `packageDirectory` for some of this.

Reply:
This looks correct to me. The reason you need a `currentDirectory` is that if you do something like `tsc --rootDir src --outDir out --declaration src/index.ts`, `options.rootDir` will be `src` and legitimately has to be resolved relative to the CWD where `tsc` was invoked.